### ISSUE TITLE
OpenAPI 3.1.0 support: changing server url

### ISSUE THREAD

**Issue by transfluxus** [swagger-api/swagger-js#2967]
### Q&A (please complete the following information)
 - OS: Ubuntu
 - Environment: firefox 113.0.1, node 18
 - Method of installation: npm
 - Swagger-Client version: 3.19.7
 - Swagger/OpenAPI version: OpenAPI 3.0

### How can we help?

I have noticed a difference in behaviour between version 3.16 and 3.19, and I am trying to replicate the result in 3.19.
Basically I want to change the server url dynamically. 
Before I would just go into the swaggerClient.spec.servers and change the array, but now this does not seem to have any effect on the following requests. How can I achieve changing the server url in the latest version?
thanks!

**Comment by char0n** [comment]
Hi @transfluxus,

Unfortunately the info you provided about the issue is very limited and not really actionable. 
I would need a full working example code + fixture where I can see what is happening. Would you be able to provide one?

**Comment by transfluxus** [comment]
Hi char0n,

thanks for getting back.
I just put together this:
```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
    <script src="https://unpkg.com/swagger-client"></script>
</head>
<body>

<script async>
    async function make_client() {
        const response = await fetch("localcontextshub_openapi.json")
        const spec = await response.json()
        return await SwaggerClient("http://localhost:9999/localcontextshub_openapi.json", {spec})
    }

    let swaggerClient;
    make_client().then(client => {
        swaggerClient = client
        console.log(swaggerClient)
        swaggerClient.execute({"operationId": "projects_list"}).then(response => {
            console.log(response)
        }, error => {
            console.error(error)
        })

        swaggerClient.spec.servers = [
            {
                "url": "https://localcontextshub.org"
            }
        ]

        swaggerClient.execute({"operationId": "projects_list"}).then(response => {
            console.log(response)
        }, error => {
            console.error(error)
        })
    })
</script>
</body>
</html>
```
which works 🥳. However in my application I am forced to use version 3.16.1
because of https://github.com/swagger-api/swagger-js/issues/2187# 
I am not sure, if that causes the problem, or if there is something else wrong in my app.. But I already tested a bit...
Anyway since it works, this ca be closed...

**Comment by char0n** [comment]
All right, closing. 

> in my application I am forced to use version 3.16.1 because of https://github.com/swagger-api/swagger-js/issues/2187

Can you elaborate on this? What in #2187 is forcing you to use 3.16.1? I've already provided fix for the regression introduced there in https://github.com/swagger-api/swagger-js/compare/v3.18.1...v3.18.2

**Comment by transfluxus** [comment]
I cannot tell now, cuz when just again updating the package to 3.19.8, it does not use the right server even if the spec looks ok
https://gist.github.com/transfluxus/46e2f6035d5f2ae2f642e53fdc569b11

It is not making the request to the right url (but the local server, since the server array is empty intially)

I went into it with the debugger and in the buildRequest function the url is not set from the spec...
In that line before the path is set, the req.url is just an empty string.
req.url += pathName; // Have not yet replaced the path parameters.

the problem with lodash was, ... that it was basically not imported, so "get" would be null and it would say something like "schema not found" because it was supposed to get the security schema with lodash get

**Comment by char0n** [comment]
@transfluxus thanks for more info. Does it only fail for `OpenAPI 3.1.0` and work for `OpenAPI 3.0.3` or it fails for both?

**Comment by char0n** [comment]
All right, thanks for investigation. So it's **OpenAPI 3.1.0 related regression**.

**Comment by char0n** [comment]
After deep investigation, here are my observations:

### Observation 1

The reason why it works for you for OpenAPI 3.0.x definition and doesn't for OpenAPI 3.1.0 definition, is that for OpenAPI 3.1.0 swagger-client uses [ApiDOM](https://github.com/swagger-api/apidom) to facilitate URI resolution and dereferencing mechanism. For OpenAPI 2.0 and OpenAPI 3.0.x internal swagger-client [specmap](https://github.com/swagger-api/swagger-js/tree/master/src/specmap) mechanism is used for URI resolution and dereferencing. `specmap` suffers with many bugs and some of these bugs are unsolvable given the current architecture.  

Part of OpenAPI 3.1.0 dereferencing is using something called normalization. For servers specifically it normalizes overrides defined by the OpenAPI 3.1.0 spec:

```
 List of Server Objects can be defined in OpenAPI 3.1 on multiple levels:

  - OpenAPI.servers
  - PathItem.servers
  - Operation.servers

 If an alternative server object is specified at the Path Item Object level, it will override OpenAPI.servers.
 If an alternative server object is specified at the Operation Object level, it will override PathItem.servers and OpenAPI.servers respectively.
```

Now legacy `specmap` doesn't do that and this is the reason why we're seeing the difference.


Created an issue with OpenAPI spec authors to clarify how `servers` fixed field should behave: https://github.com/OAI/OpenAPI-Specification/issues/3427

--- 

I'll be adding observations as I move forward with this.

**Comment by char0n** [comment]
Fix for OpenAPI 3.1.0 released in [v3.24.4](https://github.com/swagger-api/swagger-js/releases/tag/v3.24.4).

**Comment by transfluxus** [comment]
fantastic, thanks a lot. I will consider your observations and try the corresponding changes
